src/core/engines/full-margin-v2/FullMarginEngineV2.sol

```diff
@@ -0,0 +1,479 @@
+    using ArrayUtil for uint256[];
+
+    using AccountUtil for Balance[];
+    using AccountUtil for FullMarginDetailV2[];
+    using AccountUtil for Position[];
+    using AccountUtil for PositionOptim[];
+    using AccountUtil for SBalance[];
+
+    using FullMarginLibV2 for FullMarginAccountV2;
+    using FullMarginMathV2 for FullMarginDetailV2;
+    using SafeCast for uint256;
+    using SafeCast for int256;
+    using TokenIdUtil for uint256;
+
+    /*///////////////////////////////////////////////////////////////
+                                  Variables
+    //////////////////////////////////////////////////////////////*/
+
+    ///@dev subAccount => FullMarginAccountV2 structure.
+    ///     subAccount can be an address similar to the primary account, but has the last 8 bits different.
+    ///     this give every account access to 256 sub-accounts
+    mapping(address => FullMarginAccountV2) internal accounts;
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address _grappa, address _optionToken) BaseEngine(_grappa, _optionToken) {}
+
+    /*///////////////////////////////////////////////////////////////
+                        External Functions
+    //////////////////////////////////////////////////////////////*/
+
+    function batchExecute(BatchExecute[] calldata batchActions) public nonReentrant {
+        uint256 i;
+        for (i; i < batchActions.length; ) {
+            address subAccount = batchActions[i].subAccount;
+            ActionArgs[] calldata actions = batchActions[i].actions;
+
+            _execute(subAccount, actions);
+
+            // increase i without checking overflow
+            unchecked {
+                ++i;
+            }
+        }
+
+        for (i = 0; i < batchActions.length; ) {
+            if (!_isAccountAboveWater(batchActions[i].subAccount)) revert BM_AccountUnderwater();
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function execute(address _subAccount, ActionArgs[] calldata actions) public override nonReentrant {
+        _execute(_subAccount, actions);
+
+        if (!_isAccountAboveWater(_subAccount)) revert BM_AccountUnderwater();
+    }
+
+    /**
+     * @notice payout to user on settlement.
+     * @dev this can only triggered by Grappa, would only be called on settlement.
+     * @param _asset asset to transfer
+     * @param _recipient receiber
+     * @param _amount amount
+     */
+    function payCashValue(
+        address _asset,
+        address _recipient,
+        uint256 _amount
+    ) public override(BaseEngine, IMarginEngine) {
+        BaseEngine.payCashValue(_asset, _recipient, _amount);
+    }
+
+    /**
+     * @notice get minimum collateral needed for a margin account
+     * @param _subAccount account id.
+     * @return balances array of collaterals and amount (signed)
+     */
+    function getMinCollateral(address _subAccount) external view returns (SBalance[] memory balances) {
```

Comment:
What does it mean to return an amount lower than 0?
Does that mean people can have negative cash balance in the vault?